File: leetcode/63/c++/solution.cpp
/*
    63. Unique Paths II
    Author: Miguel Angel Bermeo Ayerbe
*/
class Solution {
   public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        size_t m = obstacleGrid.size(), n = obstacleGrid[0].size();

        if (!(obstacleGrid[0][0] == 0 && obstacleGrid[m - 1][n - 1] != 1)) {
            return 0;
        }
        obstacleGrid[0][0] = -1;
        for (size_t i = 0; i < m; ++i) {
            for (size_t j = 0; j < n; ++j) {
                if (obstacleGrid[i][j] != 1) {
                    if (i > 0 && obstacleGrid[i - 1][j] != 1) {
                        obstacleGrid[i][j] += obstacleGrid[i - 1][j];
                    }
                    if (j > 0 && obstacleGrid[i][j - 1] != 1) {
                        obstacleGrid[i][j] += obstacleGrid[i][j - 1];
                    }
                }
            }
        }
        return -obstacleGrid[m - 1][n - 1];
    }
};

File: leetcode/63/rust/solution.rs
/*
    63. Unique Paths II
    Author: Miguel Angel Bermeo Ayerbe
*/
impl Solution {
    pub fn unique_paths_with_obstacles(obstacle_grid: Vec<Vec<i32>>) -> i32 {
        let (n,m) = (obstacle_grid.len(),obstacle_grid[0].len());
        if obstacle_grid[0][0]+obstacle_grid[n-1][m-1] > 0{
            return 0;
        }

        let mut row_i: Vec<i32> = vec![0;m];
        
        row_i[0] = -1;
        for i in 0..n {
            for j in 0..m {
                if  obstacle_grid[i][j] == 0 {
                    if i > 0 && obstacle_grid[i-1][j] == 1 {
                        row_i[j] = 0;
                    }
                    if j > 0 && obstacle_grid[i][j-1] == 0 {
                        row_i[j] += row_i[j-1];
                    }

                }
            } 
        }
        -row_i[m-1]
    }
}